src/core/Akka.Remote/RemoteWatcher.cs

```diff
@@ -389,107 +378,116 @@ protected virtual void Quarantine(Address address, int? addressUid)
             _remoteProvider.Quarantine(address, addressUid);
         }
 
-        private void ProcessRewatchRemote(IActorRef watchee, IActorRef watcher)
-        {
-            if (_watching.Contains(Tuple.Create(watchee, watcher)))
-                ProcessWatchRemote(watchee, watcher);
-            else
-                //has been unwatched inbetween, skip re-watch
-                _log.Debug("Ignoring re-watch after being unwatched in the meantime: [{0} -> {1}]", watcher.Path,
-                    watchee.Path);
-        }
-
-        private void ProcessWatchRemote(IActorRef watchee, IActorRef watcher)
+        protected void AddWatching(IInternalActorRef watchee, IInternalActorRef watcher)
         {
-            if (watcher != Self)
+            // TODO: replace with Code Contracts assertion
+            if(watcher.Equals(Self)) throw new InvalidOperationException("Watcher cannot be the RemoteWatcher!");
+            Log.Debug("Watching: [{0} -> {1}]", watcher.Path, watchee.Path);
+            if (!Watching.ContainsKey(watchee))
             {
-                _log.Debug("Watching: [{0} -> {1}]", watcher.Path, watchee.Path);
-                AddWatching(watchee, watcher);
-
-                // also watch from self, to be able to cleanup on termination of the watchee
-                Context.Watch(watchee);
-                _watching.Add(Tuple.Create(watchee, Self));
+                Watching.Add(watchee, new HashSet<IInternalActorRef>());
             }
+            Watching[watchee].Add(watcher);
+            WatchNode(watchee);
+
+            // add watch from self, this will actually send a Watch to the target when necessary
+            Context.Watch(watchee);
         }
 
-        private void AddWatching(IActorRef watchee, IActorRef watcher)
+        protected virtual void WatchNode(IInternalActorRef watchee)
         {
-            _watching.Add(Tuple.Create(watchee, watcher));
             var watcheeAddress = watchee.Path.Address;
-            if (!_watchingNodes.Contains(watcheeAddress) && _unreachable.Contains(watcheeAddress))
+            if (!WatcheeByNodes.ContainsKey(watcheeAddress) && Unreachable.Contains(watcheeAddress))
             {
-                // first watch to that node after previous unreachable
-                _unreachable.Remove(watcheeAddress);
+                // first watch to a node after a previous unreachable
+                Unreachable.Remove(watcheeAddress);
                 _failureDetector.Remove(watcheeAddress);
             }
-            _watchingNodes.Add(watcheeAddress);
+
+            if (!WatcheeByNodes.ContainsKey(watcheeAddress))
```

Comment:
You're making unnecessary additional random access to dictionary in this section of the code. This could be rewritten to:

```csharp
HashSet<IInternalActorRef> watchees;
if (WatcheeByNodes.TryGetValue(watcheeAddress, out watchees)) 
    watchees.Add(watchee);
else WatcheeByNodes.Add(watcheeAddress, new HashSet<IInternalActorRef> { watchee });
```